src/metabase/notification/send.clj

```diff
@@ -188,10 +181,47 @@
       (prometheus/observe! :metabase-notification/total-duration-ms {:payload-type payload_type} total-time))
     nil))
 
+(defn create-notification-dispatcher
+  "Create a thread pool for sending notifications.
+  There can only be one notification with the same id in the queue.
+  - if a notification of the same id is already in the queue, then replace it
+  - if a notification doesn't have id, put it into queue regardless (used to send unsaved notifications)"
+  [pool-size]
+  (let [ids-queue        (java.util.concurrent.LinkedBlockingQueue.)
+        id->notification (atom {})
+        executor         (Executors/newFixedThreadPool
+                          pool-size
+                          (.build
+                           (doto (BasicThreadFactory$Builder.)
+                             (.namingPattern "send-notification-thread-pool-%d"))))
+        queue-lock       (Object.)]
+    (dotimes [_ pool-size]
+      (.submit ^ExecutorService executor
+               ^Callable (fn []
+                           (while true
+                             (try
+                               (when-let [id (.take ids-queue)]
+                                 (let [notification (get @id->notification id)]
+                                   (swap! id->notification dissoc id)
+                                   (send-notification-sync! notification)))
+                               (catch Exception e
+                                 (log/error e "Error in notification worker")))))))
+    (fn [notification]
+      (let [id (or (:id notification) (random-uuid))]
+        (locking queue-lock
+          (swap! id->notification assoc id notification)
+          ;; ensure there is only one notification with the same id in the queue
+          (if (.contains ids-queue id)
+            (log/infof "Notification %s is already in the queue, replacing" id)
+            (do
+              (log/infof "Adding notification %s to the queue" id)
+              (.put ids-queue id))))))))
+
+(def ^:private dispatcher
```

Comment:
I would guess this wants to be `defonce`

Reply:
defonce it goes :D